### PR TITLE
framework(engine): refactor worker exiting logic

### PR DESCRIPTION
**Pull request by liuzix**
### What problem does this PR solve?

Issue Number: ref #5805 

### What is changed and how it works?
- Implemented `NotifyExit`.
- Moved the logic controlling exiting into `MasterClient`.

#### Tests 

 - Unit test
 - Integration test

##### Will it cause performance regression or break compatibility?
No.

##### Do you need to update user documentation, design documentation or monitoring documentation?
No.

### Release note 

```release-note
None.
```

### CODE REVIEW

File: engine/framework/worker.go [hunk 1]
```diff
@@ -536,13 +539,12 @@ func (w *DefaultBaseWorker) runWatchDog(ctx context.Context) error {
...
 		}
 
-		isNormal, err := w.masterClient.CheckMasterTimeout(w.clock)
+		isNormal, err := w.masterClient.CheckMasterTimeout()
 		if err != nil {
 			return errors.Trace(err)
 		}
 		if !isNormal {
 			errOut := derror.ErrWorkerSuicide.GenWithStackByArgs(w.masterClient.MasterID())
-			w.exitController.ForceExit(errOut)
```
Comment (amyangfei) [inline engine/framework/worker.go:545]: Is `ForceExit` not used now?

Comment (liuzix) [reply engine/framework/worker.go:545]: No, it is not.

**Review by amyangfei — APPROVED** [review]

File: engine/framework/internal/worker/master_client.go [hunk 2]
```diff
@@ -214,13 +237,15 @@ func (m *MasterClient) CheckMasterTimeout(clk clock.Clock) (ok bool, err error)
 }
 
 // SendHeartBeat sends a heartbeat to the master.
-func (m *MasterClient) SendHeartBeat(ctx context.Context, clock clock.Clock, isFinished bool) error {
+func (m *MasterClient) SendHeartBeat(ctx context.Context) error {
```
Comment (amyangfei) [inline engine/framework/internal/worker/master_client.go:240]: Better to record more information in the log of L261, such as `epoch` and `sendTime`.

**Comment by amyangfei** [comment]
Please resolve conflicts @liuzix

File: engine/framework/base_jobmaster.go [hunk 3]
```diff
@@ -216,16 +214,38 @@ func (d *DefaultBaseJobMaster) GetWorkers() map[frameModel.WorkerID]WorkerHandle
...
-	// that we can close inner resources of the framework
-	if err != nil {
-		log.L().Error("Failed to close JobMasterImpl", zap.Error(err))
-	}
+	d.closeOnce.Do(func() {
+		err := d.impl.CloseImpl(ctx)
+		if err != nil {
+			log.L().Error("Failed to close JobMasterImpl", zap.Error(err))
+		}
+	})
```
Comment (maxshuang) [inline engine/framework/base_jobmaster.go:222]: If `closeOnce` and `CloseImpl` is strongly binding, may be we can wrap them together.

**Review by maxshuang — APPROVED** [review]

**Pull request merged**